### PR TITLE
Typedoc Updates

### PR DESCRIPTION
**Pull request by tnorling**
Centralizes typedoc generation into the root of the repo. This provides several benefits:
- Typedoc page provides links to all of our libs
image
- Type definitions are linked across libs
image

Also adds Typedocs for `msal-react`

New Typedocs are live [here](https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules)

### CODE REVIEW

File: lib/msal-react/package.json [hunk 1]
```diff
@@ -29,6 +29,9 @@
     "test:coverage": "tsdx test .*.spec.* --coverage",
     "lint": "cd ../../ && npm run lint:react",
     "lint:fix": "npm run lint -- -- --fix",
+    "doc": "npm run doc:generate && npm run doc:deploy",
+    "doc:generate": "typedoc src/index.ts",
+    "doc:deploy": "gh-pages -d ref -a -e ref/msal-react",
```
Comment (jasonnutter) [inline lib/msal-react/package.json:34]: Have you tried deploying?

Comment (jasonnutter) [reply lib/msal-react/package.json:34]: https://azuread.github.io/microsoft-authentication-library-for-js/ref/msal-react/

**Review by jasonnutter — CHANGES_REQUESTED** [review]
Can you add msal-react to this file? https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/3d04da6f9eea4ab49f5e33566a0871745e5aa8e4/.github/workflows/typedoc.yml

File: lib/msal-react/typedoc.json [hunk 2]
```diff
@@ -0,0 +1,5 @@
+{
+    "out": "./ref",
+    "excludePrivate": true,
+    "excludeProtected": true
+}
```
Comment (jasonnutter) [inline lib/msal-react/typedoc.json:5]: 👀 didnt know about this file. Would be good to update the other libs (can be done in separate PR).

File: lib/msal-react/package.json (hunk 1 above)
Comment (tnorling) [reply lib/msal-react/package.json:34]: Was getting some sort of permission error when I tried to do it from my machine. May need some help.

**Comment by tnorling** [comment]
> Can you add msal-react to this file? https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/3d04da6f9eea4ab49f5e33566a0871745e5aa8e4/.github/workflows/typedoc.yml

Done

File: lib/msal-react/package.json (hunk 1 above)
Comment (jasonnutter) [reply lib/msal-react/package.json:34]: You can also add this branch to the typedoc workflow file to get it to deploy these typedocs automatically.

Comment (tnorling) [reply lib/msal-react/package.json:34]: Cool. That worked!

File: lib/msal-react/typedoc.json (hunk 2 above)
Comment (tnorling) [reply lib/msal-react/typedoc.json]: Made some changes to where and how typedocs are generated. We can discuss in our sync tomorrow.

**Comment by jasonnutter** [comment]
One thing is that the homepage for the respective modules don't pull in their README:

https://azuread.github.io/microsoft-authentication-library-for-js/ref/msal-browser/
vs
https://azuread.github.io/microsoft-authentication-library-for-js/ref/msal-browser/

Any idea if that can be addressed? Not a big deal, would be nice if so.

File: lib/msal-angular/README.md [hunk 3]
```diff
@@ -5,7 +5,7 @@ Note: `msal-angular@2` is under development. **We do not recommend using this ve
 
 ![npm (scoped)](https://img.shields.io/npm/v/@azure/msal-angular)![npm](https://img.shields.io/npm/dw/@azure/msal-angular)
 
-| [Getting Started](https://docs.microsoft.com/azure/active-directory/develop/tutorial-v2-angular)| [AAD Docs](https://aka.ms/aaddevv2) | [Library Reference](https://azuread.github.io/microsoft-authentication-library-for-js/ref/msal-angular/) | [Support](README.md#community-help-and-support) | [Samples](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-angular-v2-samples)
+| [Getting Started](https://docs.microsoft.com/azure/active-directory/develop/tutorial-v2-angular)| [AAD Docs](https://aka.ms/aaddevv2) | [Library Reference](https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_angular.html) | [Support](README.md#community-help-and-support) | [Samples](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples/msal-angular-v2-samples)
```
Comment (jasonnutter) [inline lib/msal-angular/README.md:8]: Can this be made to open a new window (like with browser)?

File: lib/msal-react/README.md [hunk 4]
```diff
@@ -4,8 +4,8 @@
 
 `msal-react` is under development. **We do not recommend using this in a production environment yet**.
 
-| [AAD Docs](https://aka.ms/aaddevv2) | [Support](README.md#community-help-and-support) | [Samples](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples)
-| --- | --- | --- |
+| [AAD Docs](https://aka.ms/aaddevv2) | [Library Reference](https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_react.html) | [Support](README.md#community-help-and-support) | [Samples](https://github.com/AzureAD/microsoft-authentication-library-for-js/tree/dev/samples)
```
Comment (jasonnutter) [inline lib/msal-react/README.md:7]: Same thing here re: new window.

File: lib/msal-angular/README.md (hunk 3 above)
Comment (tnorling) [reply lib/msal-angular/README.md]: Done

File: lib/msal-react/README.md (hunk 4 above)
Comment (tnorling) [reply lib/msal-react/README.md]: Done

**Comment by tnorling** [comment]
> One thing is that the homepage for the respective modules don't pull in their README:
> 
> https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules/_azure_msal_browser.html
> vs
> https://azuread.github.io/microsoft-authentication-library-for-js/ref/msal-browser/
> 
> Any idea if that can be addressed? Not a big deal, would be nice if so.

Was able to have our root readme show up on the main [page](https://azuread.github.io/microsoft-authentication-library-for-js/ref/modules.html) but not sure at the moment how to get the individual ones for each repo. I saw there's some plugins that allow you to integrate markdown files into the typedocs. That's probably worth playing around with incrementally.

**Review by jasonnutter — APPROVED** [review]

**Review by jo-arroyo — APPROVED** [review]
LGTM!

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 5]
```diff
@@ -69,7 +69,7 @@ export abstract class ClientApplication {
      * In Azure B2C, authority is of the form https://{instance}/tfp/{tenant}/{policyName}/
      * Full B2C functionality will be available in this library in future versions.
      *
-     * @param {@link (Configuration:type)} configuration object for the MSAL PublicClientApplication instance
+     * @param configuration Object for the MSAL PublicClientApplication instance
```
Comment (sameerag) [inline lib/msal-browser/src/app/ClientApplication.ts:72]: tsdoc  tip: format apparently is supposed to be @param `parameter` - `short description`

File: lib/msal-browser/src/app/ClientApplication.ts [hunk 6]
```diff
@@ -578,7 +580,7 @@ export abstract class ClientApplication {
     /**
      * Use to log out the current user, and redirect the user to the postLogoutRedirectUri.
      * Default behaviour is to redirect the user to `window.location.href`.
-     * @param {@link (EndSessionRequest:type)}
+     * @param logoutRequest
```
Comment (sameerag) [inline lib/msal-browser/src/app/ClientApplication.ts:583]: One more inconsistency: we either add @returns for all public APIs or just don't!

**Review by sameerag — APPROVED** [review]
Approved with comments. Can be addressed with a new PR.

File: lib/msal-browser/src/app/ClientApplication.ts (hunk 6 above)
Comment (jasonnutter) [reply lib/msal-browser/src/app/ClientApplication.ts:583]: I can look at having this enforced in the js doc linting pr.

**Pull request merged**